File: elixir/lasagna/lib/lasagna.ex
defmodule Lasagna do
  @doc ~S"""
  Returns the number of expected minutes in oven.

  ## Examples

      iex> Lasagna.expected_minutes_in_oven()
      40

  """
  def expected_minutes_in_oven do
    40
  end

  @doc ~S"""
  Returns the remaining oven time in minutes.

  ## Examples

      iex> Lasagna.remaining_minutes_in_oven(30)
      10

  """
  def remaining_minutes_in_oven(current_time) do
    expected_minutes_in_oven() - current_time
  end

  @doc ~S"""
  Returns the preparation time in minutes.

  ## Examples

      iex> Lasagna.preparation_time_in_minutes(2)
      4

  """
  def preparation_time_in_minutes(num_layers) do
    2*num_layers
  end

  @doc ~S"""
  Returns the total working time in minutes.

  ## Examples

      iex> Lasagna.total_time_in_minutes(3, 20)
      26

  """
  def total_time_in_minutes(num_layers, time_in_oven) do
    preparation_time_in_minutes(num_layers) + time_in_oven
  end

  @doc ~S"""
  Returns a notification that the lasagna is ready.

  ## Examples

      iex> Lasagna.alarm()
      "Ding!"

  """
  def alarm do
    "Ding!"
  end
end

File: swift/lasagna/Sources/Lasagna/Lasagna.swift
let expectedMinutesInOven = 40

/// 
/// - Parameter elapsedMinutes: the number of minutes the lasagna has been in the oven 
/// - Returns: the number of minutes remaining based on the expected bake time
func remainingMinutesInOven(elapsedMinutes: Int) -> Int {
    return expectedMinutesInOven - elapsedMinutes
}

/// 
/// - Parameter layers: the number of layers in the lasagna
/// - Returns: the total preparation time in minutes
func preparationTimeInMinutes(layers: Int) -> Int {
    return layers * 2
}

/// 
/// - Parameters:
///   - layers: the number of layers in the lasagna
///   - elapsedMinutes: the number of minutes the lasagna has been in the oven
/// - Returns: the total time in minutes how many minutes in total you've worked on cooking 
///            the lasagna, which is the sum of the preparation time in minutes, and the time 
///            in minutes the lasagna has spent in the oven at the moment.
func totalTimeInMinutes(layers: Int, elapsedMinutes: Int) -> Int {
    return preparationTimeInMinutes(layers: layers) + elapsedMinutes
}
